Work branch

erastian opened the pull request:
move some methods from users part to auth. Still proceed with JWT auth

erastian committed:
Work branch

move some methods from users part to auth. Still proceed with JWT auth

mdziarkach commented on .env.example:
```diff
@@ -1 +1,6 @@
-SALT=10
\ No newline at end of file
+SALT=10
+TOKEN_SECRET=afasfasf
```
remote it please. You don't use it

erastian replied:
Removed

mdziarkach commented on src/app.ts:
```diff
@@ -33,9 +36,12 @@ export class App {
 	useMiddleware(): void {
 		this.app.use(express.json());
+		const authMiddleware = new AuthMiddleware(this.configService.get('TOKEN_SECRET'));
+		this.app.use(authMiddleware.execute.bind(authMiddleware));
```
const authMiddleware = new AuthMiddleware(this.configService.get('JWT_ACCESS_SECRET'));

erastian replied:
Fixed

mdziarkach commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	async refreshToken(req: Request, res: Response, next: NextFunction): Promise<void> {
+		this.ok(res, {});
```
here need to generate access and refresh tokens

erastian replied:
Done

mdziarkach commented on src/common/auth.middleware.ts:
```diff
@@ -0,0 +1,22 @@
+		try {
+			const result = jwt.verify(token.split(' ')[1], this.secret);
+			console.log(result);
+			//req.user = result;
```
put user in request

erastian replied:
Done

mdziarkach commented on src/auth/auth.controller.interface.ts:
```diff
@@ -0,0 +1,9 @@
+	registerUser: (req: Request, res: Response, next: NextFunction) => void;
+	loginUser: (req: Request, res: Response, next: NextFunction) => void;
+	logoutUser: (req: Request, res: Response, next: NextFunction) => void;
+	refreshToken: (req: Request, res: Response, next: NextFunction) => void;
```
token

erastian replied:
++

mdziarkach commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,65 @@
+			{ path: '/register', method: 'post', func: this.registerUser, middlewares: [new ValidateMiddleware(UserRegisterDto)] },
+			{ path: '/login', method: 'post', func: this.loginUser, middlewares: [new ValidateMiddleware(UserLoginDto)] },
+			{ path: '/logout', method: 'get', func: this.logoutUser },
+			{ path: '/refresh', method: 'get', func: this.refreshToken, middlewares: [] },
```
forget-password/restore-password

erastian replied:
Add placeholders. Add f() as soon as posible

mdziarkach commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,65 @@
+		try {
+			const { email, username, password } = body;
+			await this.authService.registerUser({ email, username, password });
+
```
client - alex.com/auth/activate?token=123
v
server - alex.com/api/auth/activate

erastian replied:
Done

erastian committed:
Work branch

Added mailer service.
Added user activation methods.
Replaced sqlite with mongo(sqlite was removed).
Rewrite registration methods.
Update all types in all components with new mongoes IDs(strings).
Added uuid, nodemailer dependencies with types.
Prisma was updated on latest versions.

erastian committed:
Work branch

Update .env.example file

erastian committed:
Work branch

Fixed auth middleware. WIP on login/logout f().

erastian committed:
Work branch

Add enum with roles to scheme. Add refreshToken f()

erastian committed:
Work branch

Update readme.md

erastian committed:
Work branch

Change nodemailer to templates usage

erastian committed:
Work branch

Refactor nodemailer to templates usage.

erastian committed:
Work branch

Refactor registration with activation with tokens scheme. Uninstall uuid. Change DB scheme.

mdziarkach commented:
please do not be lazy to make a detailed description for PRs

mdziarkach commented on .env.example:
```diff
@@ -1 +1,11 @@
-SALT=10
\ No newline at end of file
+SERVER_NAME="Here .dev server"
+API_URL=localhost:8000
```
we don't need it. Please remove this key

erastian replied:
Discussed and done. SERVER_NAME was removed to constants.ts file

mdziarkach commented on .env.example:
```diff
@@ -1 +1,11 @@
-SALT=10
\ No newline at end of file
+SERVER_NAME="Here .dev server"
```
For what is it need? I checked mailer service and looks like this is optional key

erastian replied:
Discussed. Still useful thing)

mdziarkach commented on .env.example:
```diff
@@ -1 +1,11 @@
\ No newline at end of file
+SERVER_NAME="Here .dev server"
+API_URL=localhost:8000
+CLIENT_URL=localhost:5000
```
You don't use it. Please remove this key

erastian replied:
API_URL has been removed. Var is replaced by CLIENT_URL, but it still leads directly to the API instead of the client url

mdziarkach commented on src/articles/articles.controller.ts:
```diff
@@ -60,7 +60,7 @@ export class ArticlesController extends BaseController implements IArticlesContr
 	async createArticle({ body }: Request, res: Response, next: NextFunction): Promise<void> {
 		try {
-			const authorID = 1; // Temp author ID. Will replace later
+			const authorID = '1a1'; // Temp author ID. Will replace later
```
you already have possibility for using real author

erastian replied:
Done

mdziarkach commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,133 @@
+
+	async activate(req: Request, res: Response, next: NextFunction): Promise<void> {
+		try {
+			const activationToken = req.params.link;
```
token should be in params

erastian replied:
already changed in next commits

mdziarkach commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,133 @@
+	async activate(req: Request, res: Response, next: NextFunction): Promise<void> {
+		try {
+			const activationToken = req.params.link;
+			await this.authService.activateUser(activationToken);
```
await this.authService.activateUser(userId);

erastian replied:
already changed in next commits

erastian committed:
Work branch

Changed activation f() and url to query params usage instead of using link params

mdziarkach commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,133 @@
+		try {
+			// const { refreshToken } = req.cookies;
+			// await this.authService.logOut(refreshToken); // Needs to improved with login func. Needs to stored tokens in DB. Mb create tokenService for that
+			res.clearCookie('refreshToken');
```
remove comments please

erastian replied:
++

mdziarkach commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,120 @@
+		});
+		const { token: activationLink } = this.getActivationToken(user);
+		const activationURL = `http://${this.configService.get(
+			'API_URL',
```
You need client_url

erastian replied:
Done

mdziarkach commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,120 @@
+		const decryptedUser = jwt.verify(
+			activationLink,
+			this.configService.get('JWT_ACTIVATION_SECRET'),
+		) as ITokenPayload;
```
you can find user in request
you auth middleware already decoded it
remove it please

erastian replied:
Refactored with new agreements

mdziarkach commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,120 @@
+		const user = await this.usersService.getUserByID(decryptedUser.id);
+
+		user.isActivated = true;
+		await this.usersService.activateUser(user);
```
id should be enough
```await this.usersService.activateUser(userId);```

erastian replied:
Done

mdziarkach commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,120 @@
+
+		const user = await this.usersService.getUserByID(decryptedUser.id);
+
+		user.isActivated = true;
```
remove it please

erastian replied:
++

mdziarkach commented on src/mailer/mailer.service.ts:
```diff
@@ -0,0 +1,63 @@
+	}
+
+	async sendActivationMail(emailFor: string, username: string, link: string): Promise<void> {
+		const _username = username.charAt(0).toUpperCase() + username.slice(1);
```
usually devs use _ for private methods and variables. But inside of function all variables are private
Let's common approach and remove _ prefix

erastian replied:
fixed

erastian committed:
Work branch

Minor fixes in mailer, auth & user services. Change in articles controller dummy user variable to user from request.

erastian committed:
Work branch

Changed secrets logic to use only one JWT secret key. Updated .env files. Moved nonTech vars from .env to constants/constants.ts file and changed some code for this. Add to the auth.middleware another one token listener method. Refactored user activation logic with new vars and methods.

mdziarkach commented on src/constants/constants.ts:
```diff
@@ -0,0 +1,4 @@
+export const CONSTANTS = {
+	PORT: 8000,
+	SERVER_NAME: 'Here .dev server',
+};
```
```js
export const PORT = 8000;
export const SERVER_NAME = 'Here .dev server',
```

erastian replied:
done

mdziarkach commented on src/auth/auth.controller.ts:
```diff
@@ -74,10 +74,11 @@ export class AuthController extends BaseController implements IAuthController {
 
 		try {
-			await this.authService.activateUser(`${activationToken}`);
+			//await this.authService.activateUser(user);
```
```js
await this.authService.activateUser(user.id);
```

erastian replied:
done

mdziarkach commented on src/auth/auth.controller.ts:
```diff
@@ -74,10 +74,11 @@ export class AuthController extends BaseController implements IAuthController {
 	async activate(req: Request, res: Response, next: NextFunction): Promise<void> {
-		const { token: activationToken } = req.query;
+		const user = req.user;
+		console.log(user);
```
remove console.log please
using debugging mode is more convenient

erastian replied:
done

mdziarkach commented on src/common/auth.middleware.ts:
```diff
@@ -8,14 +8,17 @@ export class AuthMiddleware implements IMiddleware {
 		}
 
-		try {
+		if (token.split(' ')[0] === 'Bearer') {
```
I think this conditions is not required
you will add Bearer prefix in any case
let's remove condition and leave only this string
```js
token = token.split(' ')[1];
```

erastian replied:
In req.query.token token comes without the "Bearer " prefix. In the future, when the activation request comes from the client, this behavior should be changed and this conditions check will be removed.

erastian replied:
Unified for waiting for activation token in req.body.activationToken

erastian committed:
Work branch

changed constants.ts, refactored activation method

mdziarkach commented on src/auth/auth.service.ts:
```diff
@@ -48,12 +48,12 @@ export class AuthService implements IAuthService {
-	async activateUser(decodedUser: User): Promise<void> {
-		if (!decodedUser) {
+	async activateUser(id: string): Promise<void> {
+		if (!id) {
```
```js
import { ObjectID } from "mongodb";

if (!ObjectID.isValid(id)) {...}
```

mdziarkach replied:
or maybe prisma already has it

erastian replied:
implemented

erastian committed:
Work branch

Back JWT_REFRESH_SECRET, added bson package for ID validation

mdziarkach commented on src/common/auth.middleware.ts:
```diff
@@ -8,17 +8,14 @@ export class AuthMiddleware implements IMiddleware {
 
 	execute(req: Request, res: Response, next: NextFunction): void {
-		let token = req.body.token || req.headers.authorization || req.query.token;
+		let token = req.body.token || req.headers.authorization || req.body.activationToken;
```
let token = req.body.token || req.headers.authorization

erastian replied:
done

erastian committed:
Work branch

Changed /activate from GET to POST and cut req.body.activationToken from auth.middleware
